projects/packages/sync/src/class-spawner.php

```diff
@@ -0,0 +1,91 @@
+<?php
+/**
+ * Sync Spawner.
+ *
+ * The class is responsible for spawning dedicated Sync requests.
+ *
+ * @package automattic/jetpack-sync
+ */
+
+namespace Automattic\Jetpack\Sync;
+
+use WP_Error;
+/**
+ * Class to manage Sync spawning.
+ * The purpose of this class is to provide the means to unblock Sync from running in the shutdown hook
+ * of regular requests by spawning a dedicated Sync request instead which will trigger Sync to run.
+ */
+class Spawner {
+
+	/**
+	 * Check if this request should trigger Sync to run.
+	 *
+	 * @access public
+	 *
+	 * @param Automattic\Jetpack\Sync\Queue $queue Queue object.
+	 *
+	 * @return boolean|WP_Error True if this is a POST request and jetpack_dedicated_sync_request is set, false otherwise, WP_Error if the nonce can't be verified.
+	 */
+	public static function is_dedicated_sync_request( $queue ) {
+
+		$is_dedicated_sync_request = isset( $_SERVER['REQUEST_METHOD'] ) &&
+			'POST' === $_SERVER['REQUEST_METHOD'] &&
+			isset( $_POST['jetpack_dedicated_sync_request'] );
+
+		if ( $is_dedicated_sync_request ) {
+			$is_valid = isset( $_POST['nonce'] ) &&
+				wp_verify_nonce( $_POST['nonce'], 'jetpack_sync_dedicated_request_' . $queue->id );
+			if ( ! $is_valid ) {
+				return new WP_Error( 'invalid_nonce' );
+			}
+
+			return true;
+		}
+
+		return false;
+	}
+
+	/**
+	 * Send a request to run Sync for a certain sync queue
+	 * through HTTP request that doesn't halt page loading.
+	 *
+	 * @access public
+	 *
+	 * @param Automattic\Jetpack\Sync\Queue $queue Queue object.
+	 *
+	 * @return boolean|WP_Error True if spawned, WP_Error otherwise.
+	 */
+	public static function spawn_sync( $queue ) {
+		if ( ! Settings::is_sync_spawning_enabled() ) {
+			return new WP_Error( 'sync_spawning_disabled', 'Sync spawning is disabled.' );
+		}
+
+		if ( $queue->is_locked() ) {
```

Comment:
@bisko Do you think the checks here are enough or should we maybe introduce the same checks as `do_sync_and_set_delays`? (eg importing, sender enabled and read only)

Reply:
It should be fine if the queue is unlocked, since at the moment this is the easiest way to check if an action is already running and we should wait.